content/en/docs/contributing/style-guide.md

```diff
@@ -0,0 +1,66 @@
+---
+title: "Style Guide"
+linkTitle: "Style Guide"
+weight: 80
+type: "docs"
+---
+
+cert-manager, like most Go projects, delegates almost all stylistic choices to `gofmt`,
+with `goimports` on top for organizing imports. Broadly speaking, if you set your editor to run
+`goimports` when you save a file, your code will be stylistically correct.
+
+cert-manager generally also follows the Kubernetes
+[coding conventions](https://www.kubernetes.dev/docs/guide/coding-convention/) and the Google
+[Go code review comments](https://github.com/golang/go/wiki/CodeReviewComments).
+
+## Organizing Imports
+
+Imports should be organized into 3 blocks, with each block separated by two newlines:
+
+```go
+import (
+	"stdlib"
+
+	"external"
+
+	"internal"
+)
+```
+
+An example might be the following, taken from
+[`pkg/acme/accounts/client.go`](https://github.com/jetstack/cert-manager/blob/0c71fe7795858b96cabcddabf706d997cd2fba3f/pkg/acme/accounts/client.go):
+
+```go
+
+import (
+	"crypto/rsa"
+	"crypto/tls"
+	"net"
+	"net/http"
+	"time"
+
+	acmeapi "golang.org/x/crypto/acme"
+
+	acmecl "github.com/jetstack/cert-manager/pkg/acme/client"
+	acmeutil "github.com/jetstack/cert-manager/pkg/acme/util"
+	cmacme "github.com/jetstack/cert-manager/pkg/apis/acme/v1"
+	"github.com/jetstack/cert-manager/pkg/metrics"
+	"github.com/jetstack/cert-manager/pkg/util"
+)
+```
+
+Once this manual split of standard library, external and internal imports has been made, it will be
+enforced automatically by `goimports` when executed in the future.
+
+## Go Module Compatibility
+
+It's important to note that cert-manager `v1` doesn't generally provide any guarantees that it can be
+imported and used as a module by other programs, even for publicly exported functions, variables,
+constants, interfaces and types under `pkg/`.
+
+Only the following modules can be explicitly relied upon to maintain the go module compatibility
+guarantee in cert-manager `v1`:
+
+- `pkg/apis`
+- `pkg/api`
+- `pkg/dns/util`
```

Comment:
this list may need expanding; any suggestions welcome

Reply:
I do see folks importing [/pkg/issuer/acme/dns/util](https://pkg.go.dev/github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util?tab=importedby) (which does not mean that we promise compatibility as such, but perhaps that we should)? Same with [pkg/util/pki](https://pkg.go.dev/github.com/jetstack/cert-manager/pkg/util/pki)

Reply:
Our example [dns provider webhook](https://github.com/cert-manager/webhook-example/blob/1f895be0fe2643c03838b043ce1d622f4024f791/example/example.go#L10) imports `pkg/acme/webhook`. Not sure what we should do with that as we've actively encouraged importing it.

https://github.com/cert-manager/webhook-example/blob/1f895be0fe2643c03838b043ce1d622f4024f791/example/example.go#L10

Reply:
I'd be inclined to add both to the list, but then I'd be inclined to add everything under `pkg/` to the list so I would say that :grin: 

More seriously, I think anything which we know is currently being imported by open source projects should be on this list unless there's an exceptional reason. I don't like breaking people unless it's absolutely 100% necessary.

Reply:
I take the opposite stance to be honest. Whilst I do agree we should aim towards everything being here, we've historically not performed reviews of code that I'd expect if we are aiming to make the project stable as a library.

I'd be more in favour of having an _extremely_ minimal publicly supported Go API surface to begin with, and gradually moving functions out and marking them as 'stable' on a case by case basis alongside whatever refactoring is required